<a href="https://colab.research.google.com/github/lunarways89/data-science-project/blob/main/MySkill/Data_Analysis_My_Skill_Short_Class_6_Jan_2024_Dhanu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Superstore Dataset

Import and display data

In [6]:
import pandas as pd

dataset = pd.read_csv('superstore.csv')
print(dataset)

      Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0          1  CA-2016-152156   11/8/2016  11/11/2016    Second Class   
1          2  CA-2016-152156   11/8/2016  11/11/2016    Second Class   
2          3  CA-2016-138688   6/12/2016   6/16/2016    Second Class   
3          4  US-2015-108966  10/11/2015  10/18/2015  Standard Class   
4          5  US-2015-108966  10/11/2015  10/18/2015  Standard Class   
...      ...             ...         ...         ...             ...   
9989    9990  CA-2014-110422   1/21/2014   1/23/2014    Second Class   
9990    9991  CA-2017-121258   2/26/2017    3/3/2017  Standard Class   
9991    9992  CA-2017-121258   2/26/2017    3/3/2017  Standard Class   
9992    9993  CA-2017-121258   2/26/2017    3/3/2017  Standard Class   
9993    9994  CA-2017-119914    5/4/2017    5/9/2017    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United 

In [8]:
dataset.head()

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [10]:
# Display the columns of the dataset
print("Columns in the dataset:")
print(dataset.columns)

Columns in the dataset:
Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')


# Data Structure

## List Section

In [11]:
# var list penampung nilai
ship_mode_list = dataset['Ship Mode'].unique()
state_list = dataset['State'].unique()
category_list = dataset['Category'].unique()
sub_category_list = dataset['Sub-Category'].unique()

In [12]:
# menampilkaan value list
print('Ship Mode :', ship_mode_list)
print('State :', state_list)
print('Category :', category_list)
print('Sub-Category :', sub_category_list)

Ship Mode : ['Second Class' 'Standard Class' 'First Class' 'Same Day']
State : ['Kentucky' 'California' 'Florida' 'North Carolina' 'Washington' 'Texas'
 'Wisconsin' 'Utah' 'Nebraska' 'Pennsylvania' 'Illinois' 'Minnesota'
 'Michigan' 'Delaware' 'Indiana' 'New York' 'Arizona' 'Virginia'
 'Tennessee' 'Alabama' 'South Carolina' 'Oregon' 'Colorado' 'Iowa' 'Ohio'
 'Missouri' 'Oklahoma' 'New Mexico' 'Louisiana' 'Connecticut' 'New Jersey'
 'Massachusetts' 'Georgia' 'Nevada' 'Rhode Island' 'Mississippi'
 'Arkansas' 'Montana' 'New Hampshire' 'Maryland' 'District of Columbia'
 'Kansas' 'Vermont' 'Maine' 'South Dakota' 'Idaho' 'North Dakota'
 'Wyoming' 'West Virginia']
Category : ['Furniture' 'Office Supplies' 'Technology']
Sub-Category : ['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']


In [13]:
# menampilkaan value list tanpa kurung dan petik
print('Ship Mode :', ', '.join(ship_mode_list))
print('State :',  ', '.join(state_list))
print('Category :',  ', '.join(category_list))
print('Sub-Category :',  ', '.join(sub_category_list))

Ship Mode : Second Class, Standard Class, First Class, Same Day
State : Kentucky, California, Florida, North Carolina, Washington, Texas, Wisconsin, Utah, Nebraska, Pennsylvania, Illinois, Minnesota, Michigan, Delaware, Indiana, New York, Arizona, Virginia, Tennessee, Alabama, South Carolina, Oregon, Colorado, Iowa, Ohio, Missouri, Oklahoma, New Mexico, Louisiana, Connecticut, New Jersey, Massachusetts, Georgia, Nevada, Rhode Island, Mississippi, Arkansas, Montana, New Hampshire, Maryland, District of Columbia, Kansas, Vermont, Maine, South Dakota, Idaho, North Dakota, Wyoming, West Virginia
Category : Furniture, Office Supplies, Technology
Sub-Category : Bookcases, Chairs, Labels, Tables, Storage, Furnishings, Art, Phones, Binders, Appliances, Paper, Accessories, Envelopes, Fasteners, Supplies, Machines, Copiers


In [14]:
# membuat tampilan table list

print("Ship Mode:")
print("-" * 12)  # separator line
for mode in ship_mode_list:
    print(f"| {mode:<10} |")  # Format each mode within a table-like structure
print("-" * 12)  # closing separator line

Ship Mode:
------------
| Second Class |
| Standard Class |
| First Class |
| Same Day   |
------------


## Sets Section

In [15]:
# var sets penampung nilai
ship_mode_set = set(dataset['Ship Mode'].unique())

# menampilkaan sets sets
print('Ship Mode :', ship_mode_set)

Ship Mode : {'Same Day', 'Second Class', 'First Class', 'Standard Class'}


## Tuples Section

In [45]:
tuples1 = (1, 'CA-2016-152156', '2016-11-08', '2016-11-11', 'Second Class',
            'CG-12520', 'Claire Gute', 'Consumer', 'United States', 'Henderson', 'Kentucky',
            42420, 'South', 'FUR-BO-10001798', 'Furniture', 'Bookcases',
            'Bush Somerset Collection Bookcase', 261.96, 2, 0, 41.9136)
print('Data Claire Gute : ',tuples1)

Data Claire Gute :  (1, 'CA-2016-152156', '2016-11-08', '2016-11-11', 'Second Class', 'CG-12520', 'Claire Gute', 'Consumer', 'United States', 'Henderson', 'Kentucky', 42420, 'South', 'FUR-BO-10001798', 'Furniture', 'Bookcases', 'Bush Somerset Collection Bookcase', 261.96, 2, 0, 41.9136)


## Dictionary Section

In [49]:
dict1 = {
    'Row ID': 1,
    'Order ID': 'CA-2016-152156',
    'Order Date': '2016-11-08',
    'Ship Date': '2016-11-11',
    'Ship Mode': 'Second Class',
    'Customer ID': 'CG-12520',
    'Customer Name': 'Claire Gute',
    'Segment': 'Consumer',
    'Country': 'United States',
    'City': 'Henderson',
    'State': 'Kentucky',
    'Postal Code': 42420,
    'Region': 'South',
    'Product ID': 'FUR-BO-10001798',
    'Category': 'Furniture',
    'Sub-Category': 'Bookcases',
    'Product Name': 'Bush Somerset Collection Bookcase',
    'Sales': 261.96,
    'Quantity': 2,
    'Discount': 0,
    'Profit': 41.9136
}

customer_name = dict1['Customer Name']
print(f"Data {customer_name} : {dict1}")

Data Claire Gute : {'Row ID': 1, 'Order ID': 'CA-2016-152156', 'Order Date': '2016-11-08', 'Ship Date': '2016-11-11', 'Ship Mode': 'Second Class', 'Customer ID': 'CG-12520', 'Customer Name': 'Claire Gute', 'Segment': 'Consumer', 'Country': 'United States', 'City': 'Henderson', 'State': 'Kentucky', 'Postal Code': 42420, 'Region': 'South', 'Product ID': 'FUR-BO-10001798', 'Category': 'Furniture', 'Sub-Category': 'Bookcases', 'Product Name': 'Bush Somerset Collection Bookcase', 'Sales': 261.96, 'Quantity': 2, 'Discount': 0, 'Profit': 41.9136}


# Conditional Statement

## IF Statement

In [52]:
# Menampilkan apakah pada dictionary ada Kategori Furnitur
if dict1['Category'] == 'Furniture':
    print(f"Order {dict1['Order ID']} contains Furniture.")

Order CA-2016-152156 contains Furniture.


## IF Else Statement

In [57]:
# Melakukan cek kondisi apakah data (dict1) berkategori Furniture atau bukan
if dict1['Category'] == 'Furniture':
    print(f"Pesanan {dict1['Order ID']} mengandung Furniture.")
else:
    print(f"Pesanan {dict1['Order ID']} tidak mengandung Furniture.")

# Melakukan cek kondisi apakah data (dict1) berkategori Technology atau bukan
if dict1['Category'] == 'Technology':
    print(f"Pesanan {dict1['Order ID']} mengandung Technology.")
else:
    print(f"Pesanan {dict1['Order ID']} tidak mengandung Technology.")

Pesanan CA-2016-152156 mengandung Furniture.
Pesanan CA-2016-152156 tidak mengandung Technology.


## IF Elif Else Statement

In [59]:
# Melakukan decision-making apakah data (dict1) memiliki kondisi yang
# sesuai dengan syarat category

nama_pelanggan = dict1['Customer Name']
kategori_produk = dict1['Category']
id_pesanan = dict1['Order ID']

if kategori_produk == 'Furniture':
    print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} berisi furnitur.")
elif kategori_produk == 'Office Supplies':
    print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} berisi perlengkapan kantor.")
elif kategori_produk == 'Technology':
    print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} berisi produk teknologi.")
else:
    print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} berisi kategori produk lain.")

Pesanan CA-2016-152156 oleh Claire Gute berisi furnitur.


## Nested IF Statement

In [62]:
# Melakukan decision-making apakah data (dict1) memiliki kondisi yang
# sesuai dengan kondisi dengan syarat Kategori dan Sub-Kategory nya

nama_pelanggan = dict1['Customer Name']
kategori_produk = dict1['Category']
sub_kategori_produk = dict1['Sub-Category']
id_pesanan = dict1['Order ID']
region = dict1['Region']


if kategori_produk == 'Furniture':
    if sub_kategori_produk == 'Chairs':
        print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} dari region {region} berisi kursi.")
    else:
        print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} dari region {region} berisi furnitur selain kursi.")
else:
    if kategori_produk == 'Technology':
        print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} dari region {region} berisi produk teknologi.")
    else:
        print(f"Pesanan {id_pesanan} oleh {nama_pelanggan} dari region {region} berisi produk selain furnitur dan teknologi.")

Pesanan CA-2016-152156 oleh Claire Gute dari region South berisi furnitur selain kursi.


## While Loop

In [69]:
data_loop = {'ID': [1, 2, 3], 'Nama': ['Andi', 'Budi', 'Cici'], 'Usia': [25, 30, 28]}
dataset = pd.DataFrame(data_loop)
data = pd.DataFrame(data_loop)

i = 0
while i < len(data):
    nama_pelanggan = data.loc[i, 'Nama']  # Ganti 'Customer Name' dengan 'Nama'
    usia_pelanggan = data.loc[i, 'Usia']   # Ganti 'Sales' dengan 'Usia' atau kolom lain yang sesuai

    print(f"Pelanggan {nama_pelanggan} berusia {usia_pelanggan} tahun.")

    i += 1

Pelanggan Andi berusia 25 tahun.
Pelanggan Budi berusia 30 tahun.
Pelanggan Cici berusia 28 tahun.


## For Loop

In [71]:
data_loop = {
    'ID': [1, 2, 3],
    'Customer Name': ['Andi', 'Budi', 'Cici'],
    'Sales': [100, 250, 180]
}
dataset = pd.DataFrame(data_loop)
data = pd.DataFrame(data_loop)

for index, row in data.iterrows():
    nama_pelanggan = row['Customer Name']
    total_penjualan = row['Sales']

    print(f"Pelanggan {nama_pelanggan} melakukan pembelian dengan total Rp{total_penjualan:,.2f}")

Pelanggan Andi melakukan pembelian dengan total Rp100.00
Pelanggan Budi melakukan pembelian dengan total Rp250.00
Pelanggan Cici melakukan pembelian dengan total Rp180.00


## Function Section

In [73]:
data_loop = {
    'ID': [1, 2, 3],
    'Customer Name': ['Andi', 'Budi', 'Cici'],
    'Sales': [100, 250, 180],
    'Category': ['Furniture', 'Technology', 'Office Supplies']
}
dataset = pd.DataFrame(data_loop)
data = pd.DataFrame(data_loop)


def tampilkan_info_penjualan(row):
    """
    Menampilkan informasi penjualan untuk satu baris data.

    Args:
        row: Baris data (Series Pandas) dari DataFrame.
    """
    nama_pelanggan = row['Customer Name']
    total_penjualan = row['Sales']
    kategori = row['Category']

    print(f"Pelanggan {nama_pelanggan} membeli produk kategori {kategori} dengan total Rp{total_penjualan:,.2f}")

# Menggunakan fungsi untuk setiap baris data
for index, row in data.iterrows():
    tampilkan_info_penjualan(row)

Pelanggan Andi membeli produk kategori Furniture dengan total Rp100.00
Pelanggan Budi membeli produk kategori Technology dengan total Rp250.00
Pelanggan Cici membeli produk kategori Office Supplies dengan total Rp180.00


# Data Analysis

## Performa Penjualan

In [16]:
# Penjualan Keseluruhan

# Menghitung total penjualan
total_sales = dataset['Sales'].sum()

# Memformat total sales ke Rupiah
formatted_total_sales = f"Rp{total_sales:,.0f}"

# Hasil total sales dalam Rupiah
print(f"Total Sales: {formatted_total_sales}")

Total Sales: Rp2,297,201


In [17]:
# Penjualan berdasarkan Kategori/Sub-Kategori
category_sales = dataset.groupby('Category')['Sales'].sum()
print(category_sales)
print()

category_sales = dataset.groupby('Category')['Sales'].sum()

# Format dalam Indonesian Rupiah
formatted_sales = category_sales.apply(lambda x: f"Rp{x:,.0f}")

print(formatted_sales)

Category
Furniture          741999.7953
Office Supplies    719047.0320
Technology         836154.0330
Name: Sales, dtype: float64

Category
Furniture          Rp742,000
Office Supplies    Rp719,047
Technology         Rp836,154
Name: Sales, dtype: object


In [18]:
# Penjualan berdasarkan Wilayah/Negara Bagian/Kota
region_sales = dataset.groupby('Region')['Sales'].sum()
print(region_sales)
print()

# Format dalam Rupiah
region_sales = dataset.groupby('Region')['Sales'].sum()
formatted_sales = region_sales.apply(lambda x: f"Rp{x:,.0f}")
print(formatted_sales)

Region
Central    501239.8908
East       678781.2400
South      391721.9050
West       725457.8245
Name: Sales, dtype: float64

Region
Central    Rp501,240
East       Rp678,781
South      Rp391,722
West       Rp725,458
Name: Sales, dtype: object


In [19]:
# Tren Penjualan dari Waktu ke Waktu
dataset['Order Date'] = pd.to_datetime(dataset['Order Date']) # convert 'order date' ke datetime object'
monthly_sales = dataset.groupby(dataset['Order Date'].dt.to_period('M'))['Sales'].sum()
print(monthly_sales)

Order Date
2014-01     14236.8950
2014-02      4519.8920
2014-03     55691.0090
2014-04     28295.3450
2014-05     23648.2870
2014-06     34595.1276
2014-07     33946.3930
2014-08     27909.4685
2014-09     81777.3508
2014-10     31453.3930
2014-11     78628.7167
2014-12     69545.6205
2015-01     18174.0756
2015-02     11951.4110
2015-03     38726.2520
2015-04     34195.2085
2015-05     30131.6865
2015-06     24797.2920
2015-07     28765.3250
2015-08     36898.3322
2015-09     64595.9180
2015-10     31404.9235
2015-11     75972.5635
2015-12     74919.5212
2016-01     18542.4910
2016-02     22978.8150
2016-03     51715.8750
2016-04     38750.0390
2016-05     56987.7280
2016-06     40344.5340
2016-07     39261.9630
2016-08     31115.3743
2016-09     73410.0249
2016-10     59687.7450
2016-11     79411.9658
2016-12     96999.0430
2017-01     43971.3740
2017-02     20301.1334
2017-03     58872.3528
2017-04     36521.5361
2017-05     44261.1102
2017-06     52981.7257
2017-07     45264.4160


In [20]:
# Tren Penjuaan dengan format Rupiah

# Konversi 'Order Date' ke datetime objects
dataset['Order Date'] = pd.to_datetime(dataset['Order Date'])

# Menghitung monthly sales
monthly_sales = dataset.groupby(dataset['Order Date'].dt.to_period('M'))['Sales'].sum()

# Format index (Order Date) ke "Month Year"
monthly_sales.index = monthly_sales.index.to_timestamp().strftime('%B %Y')

# Format nilai sales ke Rupiah
monthly_sales_formatted = monthly_sales.apply(lambda x: f"Rp{x:,.0f}")

print(monthly_sales_formatted)

Order Date
January 2014       Rp14,237
February 2014       Rp4,520
March 2014         Rp55,691
April 2014         Rp28,295
May 2014           Rp23,648
June 2014          Rp34,595
July 2014          Rp33,946
August 2014        Rp27,909
September 2014     Rp81,777
October 2014       Rp31,453
November 2014      Rp78,629
December 2014      Rp69,546
January 2015       Rp18,174
February 2015      Rp11,951
March 2015         Rp38,726
April 2015         Rp34,195
May 2015           Rp30,132
June 2015          Rp24,797
July 2015          Rp28,765
August 2015        Rp36,898
September 2015     Rp64,596
October 2015       Rp31,405
November 2015      Rp75,973
December 2015      Rp74,920
January 2016       Rp18,542
February 2016      Rp22,979
March 2016         Rp51,716
April 2016         Rp38,750
May 2016           Rp56,988
June 2016          Rp40,345
July 2016          Rp39,262
August 2016        Rp31,115
September 2016     Rp73,410
October 2016       Rp59,688
November 2016      Rp79,412
December 

In [21]:
# Tren Penjualan dari pesanan terbesar per bulan

#sorted_monthly_sales = monthly_sales.sort_values(ascending=False)
#print(sorted_monthly_sales)

# Mengekstrak nilai penjualan untuk pengurutan
sales_values = monthly_sales_formatted.str[2:].str.replace(',', '').astype(float)

# Urutkan berdasarkan nilai penjualan
sorted_monthly_sales = monthly_sales_formatted.iloc[sales_values.argsort()[::-1]]

print(sorted_monthly_sales)

Order Date
November 2017     Rp118,448
December 2016      Rp96,999
September 2017     Rp87,867
December 2017      Rp83,829
September 2014     Rp81,777
November 2016      Rp79,412
November 2014      Rp78,629
October 2017       Rp77,777
November 2015      Rp75,973
December 2015      Rp74,920
September 2016     Rp73,410
December 2014      Rp69,546
September 2015     Rp64,596
August 2017        Rp63,121
October 2016       Rp59,688
March 2017         Rp58,872
May 2016           Rp56,988
March 2014         Rp55,691
June 2017          Rp52,982
March 2016         Rp51,716
July 2017          Rp45,264
May 2017           Rp44,261
January 2017       Rp43,971
June 2016          Rp40,345
July 2016          Rp39,262
April 2016         Rp38,750
March 2015         Rp38,726
August 2015        Rp36,898
April 2017         Rp36,522
June 2014          Rp34,595
April 2015         Rp34,195
July 2014          Rp33,946
October 2014       Rp31,453
October 2015       Rp31,405
August 2016        Rp31,115
May 2015 

In [22]:
# Produk Terlaris

# Penjualan Keseluruhan
total_sales = dataset['Sales'].sum()

# Format total_sales to Rupiah
formatted_total_sales = f"Rp{total_sales:,.0f}"

print(f"Total Sales: {formatted_total_sales}")

Total Sales: Rp2,297,201


In [23]:
# Produk Terlaris dari Kuantitas Terjual

# Group by product name and sum quantities
product_sales_quantity = dataset.groupby('Product Name')['Quantity'].sum()

# Urutkan menurun pesanan untuk menemukan top sellers
top_sellers_quantity = product_sales_quantity.sort_values(ascending=False)

# Menampilkan produk terlaris dengan jumlah tertentu
N = 10  # Dapat diubah untuk menentukan angka dari top sellers
print(f"Top {N} Best-Selling Products by Quantity Sold:")
print(top_sellers_quantity.head(N))

Top 10 Best-Selling Products by Quantity Sold:
Product Name
Staples                                                       215
Staple envelope                                               170
Easy-staple paper                                             150
Staples in misc. colors                                        86
KI Adjustable-Height Table                                     74
Avery Non-Stick Binders                                        71
Storex Dura Pro Binders                                        71
GBC Premium Transparent Covers with Diagonal Lined Pattern     67
Situations Contoured Folding Chairs, 4/Set                     64
Staple-based wall hangings                                     62
Name: Quantity, dtype: int64


In [24]:
# Produk terlaris berdasarkan Pendapatan (Total Penjualan)

# Kelompokkan berdasarkan nama produk dan jumlah penjualan
product_sales_revenue = dataset.groupby('Product Name')['Sales'].sum()

# Urutkan dalam urutan menurun untuk menemukan penjual teratas
top_sellers_revenue = product_sales_revenue.sort_values(ascending=False)

# Menampilkan produk terlaris N teratas berdasarkan pendapatan
N = 10  # Dapat diubah untuk menentukan angka dari top sellers
print(f"Top {N} Best-Selling Products by Revenue:")
print(top_sellers_revenue.head(N))

Top 10 Best-Selling Products by Revenue:
Product Name
Canon imageCLASS 2200 Advanced Copier                                          61599.824
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind    27453.384
Cisco TelePresence System EX90 Videoconferencing Unit                          22638.480
HON 5400 Series Task Chairs for Big and Tall                                   21870.576
GBC DocuBind TL300 Electric Binding System                                     19823.479
GBC Ibimaster 500 Manual ProClick Binding System                               19024.500
Hewlett Packard LaserJet 3310 Copier                                           18839.686
HP Designjet T520 Inkjet Large Format Printer - 24" Color                      18374.895
GBC DocuBind P400 Electric Binding System                                      17965.068
High Speed Automatic Electric Letter Opener                                    17030.312
Name: Sales, dtype: float64


In [25]:
# Kelompokkan berdasarkan nama produk dan jumlah penjualan
product_sales_revenue = dataset.groupby('Product Name')['Sales'].sum()

# Urutkan dalam urutan menurun untuk menemukan penjual teratas
top_sellers_revenue = product_sales_revenue.sort_values(ascending=False)

# Format nilai sales ke Rupiah
top_sellers_revenue_rp = top_sellers_revenue.apply(lambda x: f"Rp{x:,.0f}")

# Menampilkan N produk terlaris berdasarkan pendapatan dalam Rupiah
N = 10  # Dapat diubah untuk menentukan angka dari top sellers
print(f"Top {N} Best-Selling Products by Revenue (in Rupiah):")
print(top_sellers_revenue_rp.head(N))

Top 10 Best-Selling Products by Revenue (in Rupiah):
Product Name
Canon imageCLASS 2200 Advanced Copier                                          Rp61,600
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind    Rp27,453
Cisco TelePresence System EX90 Videoconferencing Unit                          Rp22,638
HON 5400 Series Task Chairs for Big and Tall                                   Rp21,871
GBC DocuBind TL300 Electric Binding System                                     Rp19,823
GBC Ibimaster 500 Manual ProClick Binding System                               Rp19,024
Hewlett Packard LaserJet 3310 Copier                                           Rp18,840
HP Designjet T520 Inkjet Large Format Printer - 24" Color                      Rp18,375
GBC DocuBind P400 Electric Binding System                                      Rp17,965
High Speed Automatic Electric Letter Opener                                    Rp17,030
Name: Sales, dtype: object


In [26]:
# Penjualan berdasarkan Segmen Pelanggan

# Kelompokkan berdasarkan segmen pelanggan dan jumlahkan penjualan
segment_sales = dataset.groupby('Segment')['Sales'].sum()

# Memformat nilai penjualan ke Rupiah
segment_sales_rp = segment_sales.apply(lambda x: f"Rp{x:,.0f}")

# Menampilkan penjualan berdasarkan segmen pelanggan dalam Rupiah
print("Sales by Customer Segment (in Rupiah):")
print(segment_sales_rp)

Sales by Customer Segment (in Rupiah):
Segment
Consumer       Rp1,161,401
Corporate        Rp706,146
Home Office      Rp429,653
Name: Sales, dtype: object


## Analisis Pelanggan

In [27]:
# Segmentasi Pelanggan
print("Segmentasi Pelanggan:")
print(dataset['Segment'].value_counts())

Segmentasi Pelanggan:
Segment
Consumer       5191
Corporate      3020
Home Office    1783
Name: count, dtype: int64


In [28]:
# Lokasi Pelanggan (State)
print("Pelanggan berdasarkan Lokasi Pelanggan (State):")
print(dataset['State'].value_counts())

Pelanggan berdasarkan Lokasi Pelanggan (State):
State
California              2001
New York                1128
Texas                    985
Pennsylvania             587
Washington               506
Illinois                 492
Ohio                     469
Florida                  383
Michigan                 255
North Carolina           249
Arizona                  224
Virginia                 224
Georgia                  184
Tennessee                183
Colorado                 182
Indiana                  149
Kentucky                 139
Massachusetts            135
New Jersey               130
Oregon                   124
Wisconsin                110
Maryland                 105
Delaware                  96
Minnesota                 89
Connecticut               82
Oklahoma                  66
Missouri                  66
Alabama                   61
Arkansas                  60
Rhode Island              56
Utah                      53
Mississippi               53
Louisiana         

In [29]:
# Usia Pelanggan (Dengan asumsi memiliki kolom 'Usia Pelanggan')

# Jika tidak memiliki kolom usia, mungkin perlu menghitungnya dari data lain seperti tanggal lahir jika tersedia
if 'Usia Pelanggan' in dataset.columns:
    print("Usia Pelanggan:")
    print(dataset['Usia Pelanggan'].describe())
else:
    print("Informasi Usia Pelanggan tidak ditemukan dalam dataset.")

Informasi Usia Pelanggan tidak ditemukan dalam dataset.


In [30]:
# Hitung Nilai Pembelian Rata-rata
average_purchase_value = dataset['Sales'].sum() / dataset['Order ID'].nunique()

# Hitung Frekuensi Pembelian (dengan asumsi 'Tanggal Pesanan' dalam format datetime)
purchase_frequency = dataset.groupby('Customer ID')['Order Date'].nunique().mean()

# Perkirakan Umur Pelanggan (dalam bulan, Anda mungkin perlu menyesuaikannya)
# Ini adalah contoh sederhana; Anda mungkin memerlukan metode yang lebih canggih
customer_lifespan = 12 # Asumsikan rata-rata umur pelanggan 12 bulan

# Hitung CLTV
cltv = average_purchase_value * purchase_frequency * customer_lifespan

# Memformat CLTV ke Rupiah
formatted_cltv = f"Rp{cltv:,.0f}"

print(f"Perkiraan Nilai Pelanggan Seumur Hidup: {formatted_cltv}")

Perkiraan Nilai Pelanggan Seumur Hidup: Rp34,644


In [31]:
# Perputaran Pelanggan
# Churn adalah kehilangan pelanggan dari suatu bisnis/pelanggan yang berubah-ubah

import datetime

# Menentukan periode churn (misalnya, 6 bulan)
churn_period = datetime.timedelta(days=180)
churn_period_months = churn_period.days // 30  # Perkiraan bulan

# Dapatkan tanggal pembelian terbaru untuk setiap pelanggan
recency = dataset.groupby('Customer ID')['Order Date'].max()

# Hitung waktu sejak pembelian terakhir
time_since_last_purchase = datetime.datetime.now() - recency

# Identifikasi pelanggan yang berubah-ubah
churned_customers = time_since_last_purchase[time_since_last_purchase > churn_period].index

print(f"Jumlah pelanggan yang berubah: {len(churned_customers)}")
print(f"*Note: Dalam periode {churn_period_months} bulan")

Jumlah pelanggan yang berubah: 793
*Note: Dalam periode 6 bulan


In [32]:
import datetime
import pandas as pd

# Assuming 'dataset' is your DataFrame loaded from 'superstore.csv'

# Calculate Recency
recency = (datetime.datetime.now() - dataset.groupby('Customer ID')['Order Date'].max()).dt.days

# Calculate Frequency
frequency = dataset.groupby('Customer ID')['Order ID'].nunique()

# Calculate Monetary
monetary = dataset.groupby('Customer ID')['Sales'].sum()

# Get Customer Name
customer_name = dataset.groupby('Customer ID')['Customer Name'].first()

# Create RFM DataFrame with Customer Name and ID
rfm_df = pd.concat([recency, frequency, monetary, customer_name], axis=1)
rfm_df.reset_index(inplace=True)
rfm_df.columns = ['Customer ID', 'Recency', 'Frequency', 'Monetary', 'Customer Name']

# Define RFM score ranges
rfm_scores = {
    'Recency': [5, 4, 3, 2, 1],
    'Frequency': [1, 2, 3, 4, 5],
    'Monetary': [1, 2, 3, 4, 5]
}

# Function to assign RFM scores
def assign_rfm_score(value, score_ranges, metric):
    quantiles = rfm_df[metric].quantile([0.2, 0.4, 0.6, 0.8])
    if value <= quantiles[0.2]:
        return score_ranges[0]
    elif value <= quantiles[0.4]:
        return score_ranges[1]
    elif value <= quantiles[0.6]:
        return score_ranges[2]
    elif value <= quantiles[0.8]:
        return score_ranges[3]
    else:
        return score_ranges[4]

# Apply scoring and create RFMScore
for metric, scores in rfm_scores.items():
    rfm_df[f'{metric}Score'] = rfm_df[metric].apply(lambda x: assign_rfm_score(x, scores, metric))

# Gabungkan skor RFM ke dalam satu label segmen
rfm_df['RFMScore'] = rfm_df['RecencyScore'].astype(str) + rfm_df['FrequencyScore'].astype(str) + rfm_df['MonetaryScore'].astype(str)

# Mencetak segmen RFM dan jumlahnya
print("Customer Segments (RFM):")
print(rfm_df['RFMScore'].value_counts())

Customer Segments (RFM):
RFMScore
111    45
555    22
112    19
433    19
334    17
       ..
144     1
425     1
253     1
241     1
154     1
Name: count, Length: 113, dtype: int64


**Segmentasi Pelanggan dengan RTM**

*Recency* (R): Seberapa sering pelanggan melakukan pembelian (belakangan ini/recently).

*Frekuensi* (F): Seberapa sering pelanggan melakukan pembelian

*Moneter* (M): Berapa banyak uang yang dibelanjakan pelanggan.


Setiap metrik RFM diberi skor (biasanya dalam skala 1 hingga 5), di mana:

Skor *Recency* yang lebih tinggi (5) berarti pembelian yang lebih baru.

Skor *Frekuensi* yang lebih tinggi (5) berarti pembelian yang lebih sering.

Skor *Moneter* yang lebih tinggi (5) berarti pengeluaran yang lebih tinggi.

Dari contoh hasil score RTM diatas, yaitu :

RFMScore
*   155 694
*   145  53
*   135  34
*   125  12

155: Segmen ini mewakili pelanggan dengan skor Recency 1, skor Frequency 5, dan skor Monetary 5. Ini berarti pelanggan ini melakukan pembelian relatif lama tetapi telah melakukan banyak pembelian dan menghabiskan banyak uang di masa lalu.

Dari data diatas, bisa disimpulkan pelanggan dengan score **155** (berbelanja akhi-akhir ini melakukan pesanan yang sering dengan jumlah yang besar) yaitu sejumlah **694 pelanggan**.







Top 10 Customer dengan RTM Score 155 (bagus)

In [33]:
# Filter for customers with RFMScore 155
customers_155 = rfm_df[rfm_df['RFMScore'] == '155'].copy()

# Format 'Monetary' to Rupiah
customers_155.loc[:, 'MonetaryRp'] = customers_155['Monetary'].apply(lambda x: f"Rp{x:,.0f}")

# Display the top 10 customers with RFMScore 155
print("Top 10 Customers with RFMScore 155 (Monetary in Rupiah):")
print(customers_155[['Customer Name', 'Recency', 'Frequency', 'MonetaryRp']].head(10))

Top 10 Customers with RFMScore 155 (Monetary in Rupiah):
           Customer Name  Recency  Frequency MonetaryRp
34         Anna Häberlin     2975         12    Rp7,888
152          Clay Ludtke     2849         12   Rp10,881
232        Dennis Pardue     2923          9    Rp5,481
349     Jonathan Doherty     2959         11    Rp7,611
536  Maribeth Schnelling     2838         10    Rp7,444
669         Sanjit Chand     2914          9   Rp14,142


## Analisis Produk

In [34]:
# Profitabilitas Produk

# Group by product and sum profit
product_profit = dataset.groupby('Product Name')['Profit'].sum()

# Sort products by profit in descending order
most_profitable = product_profit.sort_values(ascending=False)

# Format profit to Rupiah
most_profitable = most_profitable.apply(lambda x: f"Rp{x:,.0f}")

# Display the most and least profitable products
print("Most Profitable Products (in Rupiah):")
print(most_profitable.head(10))  # Top 10 most profitable

print("\nLeast Profitable Products (in Rupiah):")
print(most_profitable.tail(10))  # Top 10 least profitable (which will be the bottom 10)

Most Profitable Products (in Rupiah):
Product Name
Canon imageCLASS 2200 Advanced Copier                                          Rp25,200
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind     Rp7,753
Hewlett Packard LaserJet 3310 Copier                                            Rp6,984
Canon PC1060 Personal Laser Copier                                              Rp4,571
HP Designjet T520 Inkjet Large Format Printer - 24" Color                       Rp4,095
Ativa V4110MDD Micro-Cut Shredder                                               Rp3,773
3D Systems Cube Printer, 2nd Generation, Magenta                                Rp3,718
Plantronics Savi W720 Multi-Device Wireless Headset System                      Rp3,696
Ibico EPK-21 Electric Binding System                                            Rp3,345
Zebra ZM400 Thermal Label Printer                                               Rp3,344
Name: Profit, dtype: object

Least Profitable Products (in Rupiah):
P

In [35]:
# Kinerja Produk

# Group by product and calculate metrics
product_performance = dataset.groupby('Product Name').agg({
    'Sales': 'sum',
    'Discount': 'mean',  # Average discount
    'Profit': 'sum'
})

# Format values to Rupiah and percentage
for col in ['Sales', 'Profit']:
    product_performance[col] = product_performance[col].apply(lambda x: f"Rp{x:,.0f}")
product_performance['Discount'] = product_performance['Discount'].apply(lambda x: f"{x:.1%}")  # Format as percentage with 1 decimal place

# Display the top 10 products' performance
print("Top 10 Product Performance (in Rupiah):")
print(product_performance.head(10))  # Display top 10

Top 10 Product Performance (in Rupiah):
                                                      Sales Discount Profit
Product Name                                                               
"While you Were Out" Message Book, One Form per...     Rp25    13.3%   Rp10
#10 Gummed Flap White Envelopes, 100/Box               Rp41    10.0%   Rp17
#10 Self-Seal White Envelopes                         Rp109     5.0%   Rp52
#10 White Business Envelopes,4 1/8 x 9 1/2            Rp489     5.7%  Rp223
#10- 4 1/8" x 9 1/2" Recycled Envelopes               Rp287     8.0%  Rp115
#10- 4 1/8" x 9 1/2" Security-Tint Envelopes          Rp147     7.5%   Rp65
#10-4 1/8" x 9 1/2" Premium Diagonal Seam Envel...    Rp176    10.0%   Rp64
#6 3/4 Gummed Flap White Envelopes                     Rp71    20.0%   Rp25
1.7 Cubic Foot Compact "Cube" Office Refrigerators  Rp2,706     6.7%  Rp579
1/4 Fold Party Design Invitations & White Envel...     Rp50    10.0%   Rp23


In [39]:
# Calculate total quantity sold for each product
product_quantity_sold = dataset.groupby('Product Name')['Quantity'].sum()

# Identify slow-moving products (e.g., products sold below a certain threshold)
slow_moving_threshold = 10  # Adjust this threshold as needed
slow_moving_products = product_quantity_sold[product_quantity_sold < slow_moving_threshold].sort_values().head(10) # Limit to top 10

# Calculate current stock levels using the 'Quantity' column
# Assuming 'Quantity' represents the total quantity ever ordered
current_stock = dataset.groupby('Product Name')['Quantity'].sum()

# Identify products with low stock levels (e.g., below a certain threshold)
low_stock_threshold = 5  # Adjust this threshold as needed
low_stock_products = current_stock[current_stock < low_stock_threshold].sort_values().head(10) # Limit to top 10

# Identify most stocked products (products with the highest quantity)
most_stocked_products = product_quantity_sold.sort_values(ascending=False).head(10)

# Display results
print("Top 10 Slow-Moving Products:")
print(slow_moving_products)

print("\nTop 10 Low Stock Products:")
print(low_stock_products)

print("\nTop 10 Most Stocked Products:")
print(most_stocked_products)

Top 10 Slow-Moving Products:
Product Name
Xerox 20                                                                     1
Global Enterprise Series Seating Low-Back Swivel/Tilt Chairs                 1
Boston 1900 Electric Pencil Sharpener                                        1
Bush Saratoga Collection 5-Shelf Bookcase, Hanover Cherry, *Special Order    1
Penpower WorldCard Pro Card Scanner                                          1
Barricks Non-Folding Utility Table with Steel Legs, Laminate Tops            2
Ultra Commercial Grade Dual Valve Door Closer                                2
PNY Rapid USB Car Charger - Black                                            2
Xerox 1969                                                                   2
Avery 5                                                                      2
Name: Quantity, dtype: int64

Top 10 Low Stock Products:
Product Name
Penpower WorldCard Pro Card Scanner                                          1
Xerox 20           

## Pengiriman dan Logistik

In [40]:
# Analyze shipping cost by shipping mode (assuming shipping cost is correlated with sales)
# This is a proxy analysis, as we don't have actual shipping cost data
avg_sales_by_mode = dataset.groupby('Ship Mode')['Sales'].mean()
avg_sales_by_mode = avg_sales_by_mode.apply(lambda x: f"Rp{x:,.0f}") # Format to Rupiah
print("Average Sales by Mode (Proxy for Shipping Cost) in Rupiah:")
print(avg_sales_by_mode)

# Identify products with highest sales (potential proxy for high shipping costs)
products_highest_sales = dataset.groupby('Product Name')['Sales'].sum().sort_values(ascending=False).head(10)
products_highest_sales = products_highest_sales.apply(lambda x: f"Rp{x:,.0f}") # Format to Rupiah
print("\nProducts with Highest Sales (Potential for High Shipping Costs) in Rupiah:")
print(products_highest_sales)

# Analyze sales distribution (proxy for shipping cost distribution)
sales_distribution = dataset['Sales'].describe()
sales_distribution = sales_distribution.apply(lambda x: f"Rp{x:,.0f}" if isinstance(x, (int, float)) else x) # Format to Rupiah
print("\nSales Distribution (Proxy for Shipping Cost Distribution) in Rupiah:")
print(sales_distribution)

Average Sales by Mode (Proxy for Shipping Cost) in Rupiah:
Ship Mode
First Class       Rp228
Same Day          Rp236
Second Class      Rp236
Standard Class    Rp228
Name: Sales, dtype: object

Products with Highest Sales (Potential for High Shipping Costs) in Rupiah:
Product Name
Canon imageCLASS 2200 Advanced Copier                                          Rp61,600
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind    Rp27,453
Cisco TelePresence System EX90 Videoconferencing Unit                          Rp22,638
HON 5400 Series Task Chairs for Big and Tall                                   Rp21,871
GBC DocuBind TL300 Electric Binding System                                     Rp19,823
GBC Ibimaster 500 Manual ProClick Binding System                               Rp19,024
Hewlett Packard LaserJet 3310 Copier                                           Rp18,840
HP Designjet T520 Inkjet Large Format Printer - 24" Color                      Rp18,375
GBC DocuBind P4